cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs

```diff
@@ -0,0 +1,202 @@
+
+instance FromJSON RootAddress where
+  parseJSON = withObject "RootAddress" $ \o ->
+                RootAddress
+                  <$> o .:  "addrs"
+                  <*> o .:? "advertise" .!= DoNotAdvertisePeer
+
+instance ToJSON RootAddress where
+  toJSON ra =
+    object
+      [ "addrs" .= addrs ra
+      , "advertise" .= advertise ra
+      ]
+
+-- | Transforms a 'RootAddress' into a pair of 'RelayAccessPoint' and its
+-- corresponding 'PeerAdvertise' value.
+--
+rootAddressToRelayAccessPoint
+  :: RootAddress
+  -> [(RelayAccessPoint, PeerAdvertise)]
+rootAddressToRelayAccessPoint RootAddress { addrs, advertise } =
+    [ (ra, advertise) | ra <- addrs ]
+
+data LocalRootPeers = LocalRootPeers
+  { localRoots :: RootAddress
+  , valency :: Int
+  } deriving (Eq, Show)
+
+instance FromJSON LocalRootPeers where
+  parseJSON = withObject "LocalRootPeers" $ \o ->
+                LocalRootPeers
+                  <$> o .: "localRoots"
+                  <*> o .: "valency"
+
+instance ToJSON LocalRootPeers where
+  toJSON lrpg =
+    object
+      [ "localRoots" .= localRoots lrpg
+      , "valency" .= valency lrpg
+      ]
+
+newtype LocalRootPeersGroups = LocalRootPeersGroups
+  { groups :: [LocalRootPeers]
+  } deriving (Eq, Show)
+
+instance FromJSON LocalRootPeersGroups where
+  parseJSON = withObject "LocalRootPeersGroups" $ \o ->
+                LocalRootPeersGroups
+                  <$> o .: "groups"
+
+instance ToJSON LocalRootPeersGroups where
+  toJSON lrpg =
+    object
+      [ "groups" .= groups lrpg
+      ]
+
+newtype PublicRootPeers = PublicRootPeers
+  { publicRoots :: RootAddress
+  } deriving (Eq, Show)
+
+instance FromJSON PublicRootPeers where
+  parseJSON = withObject "PublicRootPeers" $ \o ->
+                PublicRootPeers
+                  <$> o .: "publicRoots"
+
+instance ToJSON PublicRootPeers where
+  toJSON prp =
+    object
+      [ "publicRoots" .= publicRoots prp
+      ]
+
+data NetworkTopology = RealNodeTopology !LocalRootPeersGroups ![PublicRootPeers] !UseLedger
+  deriving (Eq, Show)
+
+instance FromJSON NetworkTopology where
+  parseJSON = withObject "NetworkTopology" $ \o ->
+                RealNodeTopology <$> (o .: "LocalRoots"                                     )
+                                 <*> (o .: "PublicRoots"                                    )
+                                 <*> (o .:? "useLedgerAfterSlot" .!= UseLedger DontUseLedger)
+
+instance ToJSON NetworkTopology where
+  toJSON top =
+    case top of
+      RealNodeTopology lrpg prp ul -> object [ "LocalRoots"         .= lrpg
+                                             , "PublicRoots"        .= prp
+                                             , "useLedgerAfterSlot" .= ul
+                                             ]
+
+-- | Read the `NetworkTopology` configuration from the specified file.
+-- While running a real protocol, this gives your node its own address and
+-- other remote peers it will attempt to connect to.
```

Comment:
This is oddly formulated and incorrect. The node doesn't learn of its own address by reading the topology file. I think that may have been true in `cardano-sl`.